File: sdk/android/tracelet-sdk/src/main/kotlin/com/ikolvi/tracelet/sdk/algorithm/TripManager.kt
package com.ikolvi.tracelet.sdk.algorithm

import uniffi.tracelet_core.TripManager as RustTripManager

/**
 * Tracks trips based on motion state transitions, delegating core logic to Rust.
 *
 * A "trip" starts when the device transitions to moving and ends when it
 * transitions to stationary. Collects start/stop locations, waypoints,
 * total distance (Haversine), and duration.
 */
class TripManager {
    /** Callback invoked when a trip ends with the full trip data map. */
    var onTripEnd: ((Map<String, Any?>) -> Unit)? = null

    /**
     * Callback invoked when a trip *starts*, with its freshly minted id (#402).
     *
     * Before #402 a trip only became observable once it was over, so anything
     * wanting to tag records with the trip they belong to had no moment to act
     * on and had to re-derive the boundary from raw motion changes.
     */
    var onTripStart: ((Map<String, Any?>) -> Unit)? = null

    private val rustTripManager = RustTripManager()

    /** Whether a trip is currently active. */
    val isTripActive: Boolean
        get() = rustTripManager.isTripActive()

    /** The active trip's id, or null between trips (#402). */
    val currentTripId: String?
        get() = rustTripManager.currentTripId()

    /**
     * Called on every motion state change.
     *
     * @param isMoving whether the device is now moving
     * @param latitude current latitude (if available)
     * @param longitude current longitude (if available)
     * @param timestamp current timestamp string or null
     */
    fun onMotionStateChanged(
        isMoving: Boolean,
        latitude: Double? = null,
        longitude: Double? = null,
        timestamp: Any? = null,
    ) {
        val nowMs = System.currentTimeMillis()
        val timestampMs = (timestamp as? Number)?.toLong() ?: nowMs

        val transition = rustTripManager.onMotionStateChanged(
            isMoving,
            latitude,
            longitude,
            timestampMs,
            nowMs
        )

        if (transition is uniffi.tracelet_core.TripTransition.Started) {
            val start = transition.start
            val startMap = mutableMapOf<String, Any?>()
            start.startLocation?.let {
                startMap["latitude"] = it.latitude
                startMap["longitude"] = it.longitude
            }
            onTripStart?.invoke(
                mapOf(
                    "tripId" to start.tripId,
                    "startedAt" to start.startedAtMs,
                    "startLocation" to startMap,
                ),
            )
            return
        }

        val tripData = (transition as? uniffi.tracelet_core.TripTransition.Ended)?.data

        if (tripData != null) {
            val startMap = mutableMapOf<String, Any?>()
            tripData.startLocation?.let {
                startMap["latitude"] = it.latitude
                startMap["longitude"] = it.longitude
            }

            val stopMap = mutableMapOf<String, Any?>()
            tripData.stopLocation?.let {
                stopMap["latitude"] = it.latitude
                stopMap["longitude"] = it.longitude
            }

            val waypointsMapList = tripData.waypoints.map { wp ->
                mapOf(
                    "latitude" to wp.latitude,
                    "longitude" to wp.longitude,
                    "timestamp" to wp.timestampMs,
                )
            }

            val outMap = mapOf<String, Any?>(
                "isMoving" to false,
                // #402: the join key shared with every location and driving
                // event recorded during this trip.
                "tripId" to tripData.tripId,
                "startedAt" to tripData.startedAtMs,
                "endedAt" to tripData.endedAtMs,
                "distance" to tripData.distanceMeters,
                "duration" to tripData.durationSeconds,
                "startLocation" to startMap,
                "stopLocation" to stopMap,
                "waypoints" to waypointsMapList,
            )

            onTripEnd?.invoke(outMap)
        }
    }

    /**
     * Called on every accepted tracking location to record waypoints.
     *
     * @param latitude location latitude
     * @param longitude location longitude
     * @param timestamp location timestamp
     */
    fun onLocationReceived(
        latitude: Double,
        longitude: Double,
        timestamp: Any? = null,
    ) {
        val nowMs = System.currentTimeMillis()
        val timestampMs = (timestamp as? Number)?.toLong() ?: nowMs
        rustTripManager.onLocationReceived(latitude, longitude, timestampMs)
    }

    /** Reset the trip manager state. */
    fun reset() {
        rustTripManager.reset()
    }
}

File: sdk/ios/Sources/TraceletSDK/algorithm/TripManager.swift
import Foundation

/// Tracks trips based on motion state transitions, delegating core logic to Rust.
public class TraceletTripManager {
    /// Callback invoked when a trip ends with the full trip data map.
    public var onTripEnd: (([String: Any?]) -> Void)?

    /// Callback invoked when a trip *starts*, with its freshly minted id (#402).
    ///
    /// Before #402 a trip only became observable once it was over, so anything
    /// wanting to tag records with the trip they belong to had no moment to act
    /// on and had to re-derive the boundary from raw motion changes.
    public var onTripStart: (([String: Any?]) -> Void)?

    private let rustTripManager = TripManager()

    /// Whether a trip is currently active.
    public var isTripActive: Bool {
        return rustTripManager.isTripActive()
    }

    /// The active trip's id, or nil between trips (#402).
    public var currentTripId: String? {
        return rustTripManager.currentTripId()
    }

    public init() {}

    public func onMotionStateChanged(
        isMoving: Bool,
        latitude: Double? = nil,
        longitude: Double? = nil,
        timestamp: Any? = nil
    ) {
        let now = Date()
        let nowMs = Int64(now.timeIntervalSince1970 * 1000)
        let timestampMs = (timestamp as? NSNumber)?.int64Value ?? nowMs

        guard let transition = rustTripManager.onMotionStateChanged(
            isMoving: isMoving,
            latitude: latitude,
            longitude: longitude,
            timestampMs: timestampMs,
            nowMs: nowMs
        ) else {
            // Not a trip boundary — moving while already moving, or stopped
            // while already stopped.
            return
        }

        if case .started(let start) = transition {
            var startMap: [String: Any?] = [:]
            if let startLoc = start.startLocation {
                startMap["latitude"] = startLoc.latitude
                startMap["longitude"] = startLoc.longitude
            }
            onTripStart?([
                "tripId": start.tripId,
                "startedAt": start.startedAtMs,
                "startLocation": startMap
            ])
            return
        }

        guard case .ended(let tripData) = transition else { return }

        var startMap: [String: Any?] = [:]
        if let startLoc = tripData.startLocation {
            startMap["latitude"] = startLoc.latitude
            startMap["longitude"] = startLoc.longitude
        }

        var stopMap: [String: Any?] = [:]
        if let stopLoc = tripData.stopLocation {
            stopMap["latitude"] = stopLoc.latitude
            stopMap["longitude"] = stopLoc.longitude
        }

        let waypoints = tripData.waypoints.map { wp in
            [
                "latitude": wp.latitude,
                "longitude": wp.longitude,
                "timestamp": wp.timestampMs
            ]
        }

        let outData: [String: Any?] = [
            "isMoving": false,
            // #402: the join key shared with every location and driving
            // event recorded during this trip.
            "tripId": tripData.tripId,
            "startedAt": tripData.startedAtMs,
            "endedAt": tripData.endedAtMs,
            "distance": tripData.distanceMeters,
            "duration": tripData.durationSeconds,
            "startLocation": startMap,
            "stopLocation": stopMap,
            "waypoints": waypoints
        ]

        onTripEnd?(outData)
    }

    public func onLocationReceived(
        latitude: Double,
        longitude: Double,
        timestamp: Any? = nil
    ) {
        let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
        let timestampMs = (timestamp as? NSNumber)?.int64Value ?? nowMs
        rustTripManager.onLocationReceived(
            latitude: latitude,
            longitude: longitude,
            timestampMs: timestampMs
        )
    }

    public func reset() {
        rustTripManager.reset()
    }
}
